lib/puma/reactor.rb

```diff
@@ -1,405 +1,121 @@
+        STDERR.puts "Error in select: #{e.message} (#{e.class})"
+        STDERR.puts e.backtrace
+        closed.each(&method(:wakeup!))
+        retry
+      else
+        raise
       end
     end
 
-    # The `calculate_sleep` sets the value that the `NIO::Selector#select` will
-    # sleep for in the main reactor loop when no sockets are being written to.
-    #
-    # The values kept in `@timeouts` are sorted so that the first timeout
-    # comes first in the array. When there are no timeouts the default timeout is used.
-    #
-    # Otherwise a sleep value is set that is the same as the amount of time it
-    # would take for the first element to time out.
-    #
-    # If that value is in the past, then a sleep value of zero is used.
-    def calculate_sleep
-      if @timeouts.empty?
-        @sleep_for = DefaultSleepFor
-      else
-        diff = @timeouts.first.value.timeout_at.to_f - Time.now.to_f
-
-        if diff < 0.0
-          @sleep_for = 0
-        else
-          @sleep_for = diff
-        end
-      end
+    # Wakeup all objects that have timed out.
+    def timeout!
+      # @ios is sorted by timeout so we can efficiently take only timed out objects.
+      @ios.take_while {|timeout_at, _| timeout_in(timeout_at) == 0}.each(&method(:wakeup!))
     end
 
-    # This method adds a connection to the reactor
-    #
-    # Typically called by `Puma::Server` the value passed in
-    # is usually a `Puma::Client` object that responds like an IO
-    # object.
-    #
-    # The main body of the reactor loop is in `run_internal` and it
-    # will sleep on `NIO::Selector#select`. When a new connection is added to the
-    # reactor it cannot be added directly to the `sockets` array, because
-    # the `NIO::Selector#select` will not be watching for it yet.
-    #
-    # Instead what needs to happen is that `NIO::Selector#select` needs to be woken up,
-    # the contents of `@input` added to the `sockets` array, and then
-    # another call to `NIO::Selector#select` needs to happen. Since the `Puma::Client`
-    # object can be read immediately, it does not block, but instead returns
-    # right away.
-    #
-    # This behavior is accomplished by writing to `@trigger` which wakes up
-    # the `NIO::Selector#select` and then there is logic to detect the value of `*`,
-    # pull the contents from `@input` and add them to the sockets array.
-    #
-    # If the object passed in has a timeout value in `timeout_at` then
-    # it is added to a `@timeouts` array. This array is then re-arranged
-    # so that the first element to timeout will be at the front of the
-    # array. Then a value to sleep for is derived in the call to `calculate_sleep`
-    def add(c)
-      @mutex.synchronize do
-        @input << c
-        @trigger << "*"
-      end
+    # Register the object with the selector and add to @ios.
+    def register(obj)
+      @ios << @selector.register(obj.last, :r).value = obj
```

Comment:
The precedence of this line makes my head hurt. I'm unsure what is being passed to <<, please break it up into multiple expressions.